src/ucp/core/ucp_worker.c

```diff
@@ -2350,3 +2350,125 @@ void ucp_worker_print_info(ucp_worker_h worker, FILE *stream)
+                                      req, UCS_CALLBACKQ_FLAG_ONESHOT, &cb_id);
+}
+
+/* Forward declaration */
+static unsigned ucp_worker_discard_uct_ep_progress(void *arg);
+
+static ucs_status_t
+ucp_worker_discard_uct_ep_pending_cb(uct_pending_req_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    return ucp_worker_discard_uct_ep_progress(req) ? UCS_OK : UCS_INPROGRESS;
+}
+
+static unsigned ucp_worker_discard_uct_ep_progress(void *arg)
+{
+    uct_worker_cb_id_t cb_id = UCS_CALLBACKQ_ID_NULL;
+    ucp_request_t *req       = (ucp_request_t*)arg;
+    uct_ep_h uct_ep          = req->send.discard_uct_ep.uct_ep;
+    ucp_worker_h worker      = req->send.discard_uct_ep.ucp_worker;
+    ucs_status_t status;
+
+    req->send.state.uct_comp.func  = ucp_worker_discard_uct_ep_flush_comp;
+    req->send.state.uct_comp.count = 1;
+
+    status = uct_ep_flush(uct_ep, req->send.discard_uct_ep.ep_flush_flags,
+                          &req->send.state.uct_comp);
+    if (status == UCS_ERR_NO_RESOURCE) {
+        req->send.uct.func = ucp_worker_discard_uct_ep_pending_cb;
+
+        status = uct_ep_pending_add(uct_ep, &req->send.uct, 0);
+        if (status == UCS_ERR_BUSY) {
+            /* try again to flush UCT EP */
+            uct_worker_progress_register_safe(worker->uct,
+                                              ucp_worker_discard_uct_ep_progress,
+                                              req, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                              &cb_id);
+            return 0;
+        }
+        ucs_assert(status == UCS_OK);
+    } else if (status != UCS_INPROGRESS) {
+        ucp_worker_discard_uct_ep_flush_comp(&req->send.state.uct_comp, status);
+    }
+
+    return 1;
+}
+
+void ucp_worker_discard_uct_ep(ucp_worker_h worker, uct_ep_h uct_ep,
+                               unsigned ep_flush_flags)
+{
+    uct_worker_cb_id_t cb_id = UCS_CALLBACKQ_ID_NULL;
+    ucp_wireup_ep_t *wireup_ep;
+    ucp_request_t *req;
+    int is_owner;
+
+    ucs_assert(uct_ep != NULL);
+
+    if (ucp_wireup_ep_test(uct_ep)) {
+        wireup_ep = ucp_wireup_ep(uct_ep);
+        ucs_assert(wireup_ep != NULL);
+
+        is_owner = wireup_ep->super.is_owner;
+        uct_ep   = ucp_wireup_ep_extract_next_ep(uct_ep);
+
+        /* destroy WIREUP EP allocated for this UCT EP, since
+         * discard operation most likely won't have an access to
+         * UCP EP as it could be destroyed by the caller */
+        uct_ep_destroy(&wireup_ep->super.super);
+
+        if (!is_owner) {
+            /* do nothing, if this wireup EP is not an owner for UCT EP */
+            return;
+        }
+    }
+
+    req = ucp_request_get(worker);
+    if (ucs_unlikely(req == NULL)) {
+        ucs_error("unable to allocate request for discarding UCT EP %p "
+                  "on UCP worker %p", uct_ep, worker);
+        return;
+    }
+
+    ++worker->flush_ops_count;
```

Comment:
should be in UCS_ASYNC_BLOCK because wireup process can access it from async context